runtime/eventring_consumer.c

```diff
@@ -0,0 +1,683 @@
+
+  if (cursor != NULL) {
+    caml_eventring_free_cursor(cursor);
+
+    Cursor_val(v) = NULL;
+  }
+}
+
+static int ml_runtime_begin(int domain_id, void *callback_data,
+                             uint64_t timestamp, ev_runtime_phase phase) {
+  CAMLparam0();
+  CAMLlocal4(tmp_callback, ts_val, msg_type, callbacks_root);
+
+  callbacks_root = *((value *)callback_data);
+
+  tmp_callback = Field(callbacks_root, 0); /* ev_runtime_begin */
+  if (Is_some(tmp_callback)) {
+    ts_val = caml_copy_int64(timestamp);
+    msg_type = Val_long(phase);
+
+    caml_callback3(Some_val(tmp_callback), Val_long(domain_id), ts_val,
+                   msg_type);
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static int ml_runtime_end(int domain_id, void *callback_data,
+                           uint64_t timestamp, ev_runtime_phase phase) {
+  CAMLparam0();
+  CAMLlocal4(tmp_callback, ts_val, msg_type, callbacks_root);
+
+  callbacks_root = *((value *)callback_data);
+
+  tmp_callback = Field(callbacks_root, 1); /* ev_runtime_end */
+  if (Is_some(tmp_callback)) {
+    ts_val = caml_copy_int64(timestamp);
+    msg_type = Val_long(phase);
+
+    caml_callback3(Some_val(tmp_callback), Val_long(domain_id), ts_val,
+                   msg_type);
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static int ml_runtime_counter(int domain_id, void *callback_data,
+                               uint64_t timestamp, ev_runtime_counter counter,
+                               uint64_t val) {
+  CAMLparam0();
+  CAMLlocal2(tmp_callback, callbacks_root);
+  CAMLlocalN(params, 4);
+
+  callbacks_root = *((value *)callback_data);
+
+  tmp_callback = Field(callbacks_root, 2); /* ev_runtime_counter */
+  if (Is_some(tmp_callback)) {
+    params[0] = Val_long(domain_id);
+    params[1] = caml_copy_int64(timestamp);
+    params[2] = Val_long(counter);
+    params[3] = Val_long(val);
+
+    caml_callbackN(Some_val(tmp_callback), 4, params);
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static int ml_alloc(int domain_id, void *callback_data, uint64_t timestamp,
+                     uint64_t *sz) {
+  CAMLparam0();
+  CAMLlocal4(tmp_callback, ts_val, misc_val, callbacks_root);
+
+  callbacks_root = *((value *)callback_data);
+
+  tmp_callback = Field(callbacks_root, 3); /* ev_alloc */
+  if (Is_some(tmp_callback)) {
+    int i;
+
+    ts_val = caml_copy_int64(timestamp);
+    misc_val = caml_alloc(EVENTRING_NUM_ALLOC_BUCKETS, 0);
+
+    for (i = 0; i < EVENTRING_NUM_ALLOC_BUCKETS; i++) {
+      Store_field(misc_val, i, Val_long(sz[i]));
+    }
+
+    caml_callback3(Some_val(tmp_callback), Val_long(domain_id), ts_val,
+                   misc_val);
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static int ml_lifecycle(int domain_id, void *callback_data, int64_t timestamp,
+                         ev_lifecycle lifecycle, int64_t data) {
+  CAMLparam0();
+  CAMLlocal2(tmp_callback, callbacks_root);
+  CAMLlocalN(params, 4);
+
+  callbacks_root = *((value *)callback_data);
```

Comment:
I'd like to take a look if there's any problem with the various `callback_data` bits in this (and related) functions and if there's any potential for it to be harmfully moved.